djongo/sql2mongo/converters.py

```diff
@@ -77,32 +78,62 @@ def to_mongo(self):
         return [{'$project': project}]
 
     def _using_group_by(self):
-        group = {
-            '_id': None
-        }
-        project = {
-            '_id': False
-        }
+        group = {'_id': None}
+        project = {'_id': False}
+        has_agg_distinct = any(getattr(selected, 'is_agg_distinct', False) for selected in self.sql_tokens)
+
+        ## FIX: agg(distinct) handler
+        pipeline = self._handle_agg_distinct(group, project) if has_agg_distinct \
+                            else self._handle_agg(group, project)
+
+        return pipeline
+
+    def _get_alias(self, token):
+        return token.alias or str(token.__hash__())
+
+    def _handle_agg_distinct(self, group, project):
+        pipelines, keys = [], []
+        for selected in self.sql_tokens:
+            if isinstance(selected, SQLFunc) and getattr(selected, 'is_agg_distinct', False):
+                # token = agg(distinct col)
+                key = self._get_alias(selected)
+                field = f'${selected.column}' if selected.column == selected.table else f'${selected.field}'
+                group1 = {'_id': field, key: selected.to_mongo()}
+                group2 = {'_id': None, key: ast.literal_eval(str(selected.to_mongo()).replace(field, f'${key}'))}
+                pipeline = [{'$group': group1}, {'$group': group2}]
+            elif isinstance(selected, SQLFunc) and not getattr(selected, 'is_agg_distinct', False):
+                # token = agg(col)
+                key = self._get_alias(selected)
+                group[key] = selected.to_mongo()
+                pipeline = [{'$group': group}]
+            else: # token = col
+                key = selected.field
+                pipeline = [{'$group': group}]
+            project[key] = True
+            keys.append(key)
+            pipelines.append(pipeline)
+
+        if len(pipelines) > 1: 
+            # use $facet when there are multiple selected columns
+            facet, project = {}, {}
+            for idx, pipeline in enumerate(pipelines):
+                facet[str(idx)] = pipeline
+                project[keys[idx]] = {'$arrayElemAt': [ f"${idx}.{keys[idx]}", 0]}
+            final_pipeline = [{'$facet': facet}, {'$project': project}]
+        else:
+            final_pipeline = pipelines[0] + [{'$project': project}]
```

Comment:
This code would look a lot cleaner if you just do the negative case first:
if len(pipelines) == 1:
    return pipelines[0] + [{'$project': project}]
# use $facet when there are multiple selected columns
facet, project = {}, {}
for idx, pipeline in enumerate(pipelines):
    facet[str(idx)] = pipeline
    project[keys[idx]] = {'$arrayElemAt': [ f"${idx}.{keys[idx]}", 0]}
return [{'$facet': facet}, {'$project': project}]